src/sync/impl/sync_file.cpp

```diff
@@ -0,0 +1,316 @@
+namespace realm {
+
+namespace {
+
+uint8_t value_of_hex_digit(char hex_digit)
+{
+    if (hex_digit >= '0' && hex_digit <= '9') {
+        return hex_digit - '0';
+    } else if (hex_digit >= 'A' && hex_digit <= 'F') {
+        return 10 + hex_digit - 'A';
+    } else if (hex_digit >= 'a' && hex_digit <= 'f') {
+        return 10 + hex_digit - 'a';
+    } else {
+        throw std::invalid_argument("Cannot get the value of a character that isn't a hex digit.");
+    }
+}
+
+bool filename_is_reserved(const std::string& filename) {
+    return (filename == "." || filename == "..");
+}
+
+bool character_is_unreserved(char character)
+{
+    bool is_capital_letter = (character >= 'A' && character <= 'Z');
+    bool is_lowercase_letter = (character >= 'a' && character <= 'z');
+    bool is_number = (character >= '0' && character <= '9');
+    bool is_allowed_symbol = (character == '-' || character == '_' || character == '.');
+    return is_capital_letter || is_lowercase_letter || is_number || is_allowed_symbol;
+}
+
+char decoded_char_for(const std::string& percent_encoding, size_t index)
+{
+    if (index+2 >= percent_encoding.length()) {
+        throw std::invalid_argument("Malformed string: not enough characters after '%' before end of string.");
+    }
+    REALM_ASSERT(percent_encoding[index] == '%');
+    return (16*value_of_hex_digit(percent_encoding[index + 1])) + value_of_hex_digit(percent_encoding[index + 2]);
+}
+
+} // (anonymous namespace)
+
+namespace util {
+
+void remove_nonempty_dir(const std::string& path)
+{
+#ifdef _WIN32
+    static_assert(false, "Not implemented for Win32 yet");
+#else
+    // Open the directory and list all the files.
+    DIR *dir_listing = opendir(path.c_str());
+    if (!dir_listing) {
+        return;
+    }
+    while (struct dirent *file = readdir(dir_listing)) {
+        auto file_type = file->d_type;
+        std::string file_name = file->d_name;
+        if (file_name == "." || file_name == "..") {
+            continue;
+        }
+        if (file_type == DT_REG || file_type == DT_FIFO) {
+            try {
+                File::remove(file_path_by_appending_component(path, file_name));
+            }
+            catch (File::NotFound) {
+            }
+        } else if (file_type == DT_DIR) {
+            // Directory, recurse
+            remove_nonempty_dir(file_path_by_appending_component(path, file_name, FilePathType::Directory));
+        }
+    }
+    closedir(dir_listing);
+    // Delete the directory itself
+    try {
+        util::remove_dir(path);    
+    }
+    catch (File::NotFound) {
+    }
+#endif
+}
+
+std::string make_percent_encoded_string(const std::string& raw_string)
+{
+    std::string buffer;
+    buffer.reserve(raw_string.length());
+    for (char character : raw_string) {
+        if (character_is_unreserved(character)) {
+            buffer.push_back(character);
+        } else {
+            char b[4] = {0, 0, 0, 0};
+            snprintf(b, 4, "%%%2X", character);
```

Comment:
Thanks for the explanation, knowing that `snprintf()` doesn't terminate the string if the buffer size is wrong is pretty important.